src/libs/Navigation/helpers/navigateToWorkspacesPage.ts

```diff
@@ -0,0 +1,70 @@
+import interceptAnonymousUser from '@libs/interceptAnonymousUser';
+import {getPreservedNavigatorState} from '@libs/Navigation/AppNavigator/createSplitNavigator/usePreserveNavigatorState';
+import Navigation from '@libs/Navigation/Navigation';
+import navigationRef from '@libs/Navigation/navigationRef';
+import type {WorkspaceSplitNavigatorParamList} from '@libs/Navigation/types';
+import {getPolicy, isPendingDeletePolicy, shouldShowPolicy as shouldShowPolicyUtil} from '@libs/PolicyUtils';
+import CONST from '@src/CONST';
+import NAVIGATORS from '@src/NAVIGATORS';
+import ROUTES from '@src/ROUTES';
+import SCREENS from '@src/SCREENS';
+import {isFullScreenName, isWorkspacesTabScreenName} from './isNavigatorName';
+import {getLastVisitedWorkspaceTabScreen, getWorkspacesTabStateFromSessionStorage} from './lastVisitedTabPathUtils';
+
+type Params = {
+    currentUserLogin?: string;
+    shouldUseNarrowLayout: boolean;
+};
+
+const navigateToWorkspacesPage = ({currentUserLogin, shouldUseNarrowLayout}: Params) => {
+    const rootState = navigationRef.getRootState();
+    const topmostFullScreenRoute = rootState.routes.findLast((route) => isFullScreenName(route.name));
+    if (!topmostFullScreenRoute) {
+        return;
+    }
+
+    if (topmostFullScreenRoute.name === NAVIGATORS.WORKSPACE_SPLIT_NAVIGATOR) {
+        Navigation.goBack(ROUTES.WORKSPACES_LIST.route);
+        return;
+    }
+
+    interceptAnonymousUser(() => {
+        const workspacesTabStateFromSessionStorage = getWorkspacesTabStateFromSessionStorage() ?? rootState;
+        const lastWorkspacesTabNavigatorRoute = workspacesTabStateFromSessionStorage.routes.findLast((route) => isWorkspacesTabScreenName(route.name));
+        // If there is no settings or workspace navigator route, then we should open the settings navigator.
+        if (!lastWorkspacesTabNavigatorRoute) {
+            Navigation.navigate(ROUTES.WORKSPACES_LIST.route);
+            return;
+        }
+
+        let workspacesTabState = lastWorkspacesTabNavigatorRoute.state;
+        if (!workspacesTabState && lastWorkspacesTabNavigatorRoute.key) {
+            workspacesTabState = getPreservedNavigatorState(lastWorkspacesTabNavigatorRoute.key);
+        }
+
+        // If there is a workspace navigator route, then we should open the workspace initial screen as it should be "remembered".
+        if (lastWorkspacesTabNavigatorRoute.name === NAVIGATORS.WORKSPACE_SPLIT_NAVIGATOR) {
+            const params = workspacesTabState?.routes.at(0)?.params as WorkspaceSplitNavigatorParamList[typeof SCREENS.WORKSPACE.INITIAL];
+            // Screens of this navigator should always have policyID
+            const policy = getPolicy(params.policyID);
```

Comment:
Oh, whoops, I wrote a whole comment and then it somehow got deleted. Anyways, here it is again. The latest that I've heard is that we should not be using utilities like this to get an Onyx value from a variable set by Onyx.connect. Instead, we should access the value with useOnyx in a component and then pass it to the necessary function. In order to do this, we will need to connect to all policies in the navigation tab bar and use a selector that gets the policy ID from the navigation state using the logic that we have in this function.

Does that sound reasonable to you @tgolen? I know you have been involved in this guidance.

Reply:
Yep, generally speaking, that's exactly what should happen. In fact, `getPolicy()` was recently deprecated, so I'm surprised there is not an ESLint error for this (maybe there is and I just didn't look).